### PR TITLE
DDR example

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by andresovela** [comment]
Hi @panickal-xmos, I tried to run this example but it doesn't build for me:

```
Creating app_device.xe
model.tflite.cpp: Warning: Undefined reference to '_ZN6tflite3ops5micro5xcore18Register_XC_concatEv'
Constraint check for tile[0]:
  Memory available:       524288,   used:       8636 .  OKAY
    (Stack: 524, Code: 6860, Data: 1252)
  ExtMem available:    134217728,   used:    3587236 .  OKAY
    (Stack: 0, Code: 0, Data: 3587236)
Constraints checks PASSED WITH CAVEATS.
model.tflite.cpp: Error: Undefined reference to '_ZN6tflite3ops5micro5xcore18Register_XC_concatEv'
xmake[1]: *** [bin//app_device.xe] Error 1
xmake: *** [bin//app_device.xe] Error 2
```

I have the same issue with one of my newer models but I thought maybe it was a problem with the model. Now that I see it in the example I think maybe something is missing in my setup?

**Comment by ghost** [comment]
Could you please try again with the latest tools? Install them with
`pip install xmos-ai-tools --pre --upgrade`

**Comment by andresovela** [comment]
Okay that worked, thanks! I'll try adapting my application to match the example and I'll come back to you.

File: examples/app_single_model_on_one_tile_and_DDR/src/main.xc [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+//     XS1_CLKBLK_2
+// };
+
+#define NUMBER_OF_MODELS 1
+extern void model_init(chanend f);
+extern void inference();
+
+// Hack for xc to be happy with EXTMEM
+// Have to specify the extern array size
+extern const int8_t tile_server_weights[3587236];
```
Comment (andresovela) [inline examples/app_single_model_on_one_tile_and_DDR/src/main.xc]: @panickal-xmos how did you get the array size? It's not defined in the generated `model_weights.c` file

File: examples/app_single_model_on_one_tile_and_DDR/generate_optimized_cpp_for_xcore.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    OPTIMIZED_MODEL_PATH,
+    {
+        "xcore-thread-count": "5",
+        # set conv err threshold
+        "xcore-conv-err-threshold": "3",
+        # operation splitting to reduce tensor arena size
+        "xcore-op-split-tensor-arena": "True",
+        "xcore-op-split-top-op": "0,7",
+        "xcore-op-split-bottom-op": "6,14",
+        "xcore-op-split-num-splits": "8,4",
```
Comment (andresovela) [inline examples/app_single_model_on_one_tile_and_DDR/generate_optimized_cpp_for_xcore.py:21]: Could you explain the meaning of these numbers?

File: examples/app_single_model_on_one_tile_and_DDR/generate_optimized_cpp_for_xcore.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        # write weights as an array to be 
+        # placed on second tile
+        "xcore-load-tile" : "True",
```
Comment (andresovela) [inline examples/app_single_model_on_one_tile_and_DDR/generate_optimized_cpp_for_xcore.py]: Why is this necessary to load weights from DDR?

File: examples/app_single_model_on_one_tile_and_DDR/src/support.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  }
+
+  printf("\nClass with max1 value = %d and probability = %f", maxIndex1, dequantize_output(max1));
+  printf("\nClass with max2 value = %d and probability = %f", maxIndex2, dequantize_output(max2));
+  printf("\nClass with max3 value = %d and probability = %f", maxIndex3, dequantize_output(max3));
+}
+
+extern "C" {
+void model_init(unsigned flash_data) { init(flash_data); }
```
Comment (andresovela) [inline examples/app_single_model_on_one_tile_and_DDR/src/support.cpp]: I guess this is a misnamed variable now

**Review by andresovela — COMMENTED** [review]
Added some comments just for clarification

File: examples/app_single_model_on_one_tile_and_DDR/generate_optimized_cpp_for_xcore.py (hunk 2 above)
Comment (ghost) [reply examples/app_single_model_on_one_tile_and_DDR/generate_optimized_cpp_for_xcore.py:21]: It informs the compiler to split the graph to reduce sizes of simultaneously live tensors to reduce tensor arena. In this case, we are splitting op 0 to 6, 8 times, and ops 7 to 14, 4 times. Please have a look through the attached document, which explains the numbers in more detail.
[operator_splitting_note.pdf](https://github.com/user-attachments/files/16332459/operator_splitting_note.pdf)

File: examples/app_single_model_on_one_tile_and_DDR/src/main.xc (hunk 1 above)
Comment (ghost) [reply examples/app_single_model_on_one_tile_and_DDR/src/main.xc]: I need to figure a clean way to handle this. 

While writing the code, I had temporarily added 
`char (*a)[sizeof( tile_server_weights )] = 1` in model_weights.c to force a compiler error. That gives this error, 
`../src/model_weights.c:44296:8: warning: incompatible integer to pointer conversion initializing 'char (*)[3587236]' with an expression of type 'int' [-Wint-conversion]`

File: examples/app_single_model_on_one_tile_and_DDR/generate_optimized_cpp_for_xcore.py (hunk 3 above)
Comment (ghost) [reply examples/app_single_model_on_one_tile_and_DDR/generate_optimized_cpp_for_xcore.py]: The flag `--xcore-load-tile` just informs the compiler to write out weights as a header file or c file. It was meant to be used to load weights from the second tile. It should be renamed to something like, `--xcore-write-weights-as-header-file` or so.

File: examples/app_single_model_on_one_tile_and_DDR/src/main.xc (hunk 1 above)
Comment (andresovela) [reply examples/app_single_model_on_one_tile_and_DDR/src/main.xc]: Cool trick, I'll try that haha

File: examples/app_single_model_on_one_tile_and_DDR/src/support.cpp (hunk 4 above)
Comment (ghost) [reply examples/app_single_model_on_one_tile_and_DDR/src/support.cpp]: Yeah, you're right. it uses the same channel to communicate, whether it is from flash or DDR or other tile. We could call it `weights_data` instead to be more generic.

File: examples/app_single_model_on_one_tile_and_DDR/generate_optimized_cpp_for_xcore.py (hunk 3 above)
Comment (andresovela) [reply examples/app_single_model_on_one_tile_and_DDR/generate_optimized_cpp_for_xcore.py]: Oh I see, I was wondering how that worked. My first thought was that it generated different formats depending on the file extension of `--xcore-weights-file`. Good to know.

Comment (andresovela) [reply examples/app_single_model_on_one_tile_and_DDR/generate_optimized_cpp_for_xcore.py]: I created #915 to improve this and to reduce the amount of manual tweaking after generating the files.

File: examples/app_single_model_on_one_tile_and_DDR/src/XCORE-AI-EXPLORER-800.xn [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  </Declarations>
+
+  <Packages>
+    <Package id="0" Type="XS3-UnA-1024-FB265">
+      <Nodes>
+        <Node Id="0" InPackageId="0" Type="XS3-L16A-1024" Oscillator="24MHz" SystemFrequency="800MHz" ReferenceFrequency="100MHz">
+          <Boot>
+            <Source Location="bootFlash"/>
+          </Boot>
+          <Extmem sizeMbit="1024" Frequency="200MHz">
```
Comment (andresovela) [inline examples/app_single_model_on_one_tile_and_DDR/src/XCORE-AI-EXPLORER-800.xn]: Is it enough to set the frequency to 200 MHz here? In the [LPDDR docs](https://www.xmos.com/documentation/XM-014363-PC-9/html/tools-guide/tutorials/lpddr.html) there's a mention of a slightly more complex setup:

```
<Node Id="0" InPackageId="0" Type="XS3-L16A-1024" Oscillator="24MHz"
 SystemFrequency="600MHz" ReferenceFrequency="100MHz"
 SecondaryPllInputDiv="1" SecondaryPllOutputDiv="3" SecondaryPllFeedbackDiv="83">

  <Extmem SizeMbit="1024" SourcePll="SecondaryPll" Divider="2">

    <!-- Attributes for Padctrl and Lpddr XML elements are as per equivalently named 'Node Configuration' registers in datasheet -->
    <Padctrl clk="0x30" cke="0x30" cs_n="0x30" we_n="0x30" cas_n="0x30" ras_n="0x30" addr="0x30" ba="0x30" dq="0x31" dqs="0x31" dm="0x30"/>

    <Lpddr emr_opcode="0x20"/>
  </Extmem>
```

And that's for 166 MHz.
I actually have a [thread](https://www.xcore.com/viewtopic.php?f=56&t=8806&p=41259#p41259) in the community forum asking about setting the LPDDR clock frequency to 200 MHz, but I haven't received an answer yet. Is it possible/safe to set the frequency to 200 MHz?

File: examples/app_single_model_on_one_tile_and_DDR/src/main.xc [hunk 6]
```diff
@@ -0,0 +1,27 @@
+#include <stdint.h>
+#include <stdio.h>
+#include <platform.h>
+
+#define NUMBER_OF_MODELS 1
```
Comment (andresovela) [inline examples/app_single_model_on_one_tile_and_DDR/src/main.xc]: This looks unused
